### ISSUE TITLE
Angular interceptor should support APIs with optional authentication

### ISSUE THREAD

**Issue by kevincathcart-cas** [AzureAD/microsoft-authentication-library-for-js#3758]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Description

There should be a way in the angular interceptor to specify an API where authentication is optional
(i.e. it should send the access token if it has a valid one, or if it can silent acquire one, but omit it otherwise). This is useful for APIs that do not strictly require authentication, but provide additional functionality if authenticated. It might return only public data to an anonymous user, but include personalized data for a signed in user, or signed in users may have less restrictive rate limiting, etc.

Furthermore, the need for authentication can vary for the same URL by http method, and right now there is no way to support that without a full.

For the first scenario, this feels like it should be a relatively easy feature to add, since it is simply skipping the `this.acquireTokenInteractively` calls. Even for the second scenario, sending the auth token (when available) when not actually required is usually harmless, so both scenarios could be supported in this same manner.

The only tricky part seems to me to be the ability to specify this option. I would simply update the `ProtectedResourceScopes` type to make httpMethod optional (with semantics that not supplied means any method), and add a new boolean to trigger the new behavior. The reason I would put it on the scopes level rather than the resource level is because that is where the httpMethod is specified, and this should allow (but not require) distinguishing by method.

Thoughts?

### Source

External (Customer)

**Comment by jasonnutter** [comment]
> Furthermore, the need for authentication can vary for the same URL by http method, and right now there is no way to support that without a full.

Note, this is currently possible in v2. See more here: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-angular/docs/v2-docs/msal-interceptor.md#protected-resource-map

**Comment by samuelkubai** [comment]
@kevincathcart-cas thank you for filling this issue, we'll mark it as an enhancement and put it in the backlog. We'll update this issue with any updates.

**Comment by kevincathcart-cas** [comment]
> > Furthermore, the need for authentication can vary for the same URL by http method, and right now there is no way to support that without a full.
> 
> Note, this is currently possible in v2. 

Right, I misread the source, and if specify only `ProtectedResourceScopes`, but don't include a specific httpMethod in any, it will fall back to not using auth. Good.

Obviously the point about optional authentication for a route still stands.

**Comment by hansmaad** [comment]
I think it's an important enhancement. We're currently using our own interceptor based on MsalInterceptor, because it doesn't support this scenario. What I did is 
 - Add an `optional` flag to `ProtectedResourceScopes`:
```
protectedResourceMap.set(api, [{
    httpMethod: undefined,
    scopes: ['myscope'],
    optional: true,
}]);
```
- In the interceptor `matchScopesToEndpoint` method
  - Allow `httpMethod` to be undefined -> match all methods
  - Return `{ scope: string, optional?: boolean} []` instead of `string[]`

- In `intercept` add
```
if (!account && scopes.every(s => s.optional)) {
    this.authService.getLogger().verbose('Interceptor - no account & no required scope for endpoint');
    return next.handle(req);
}
```
Is there any activity on this topic? I'd really like to get rid of my custom interceptor as it just copies 99% of the original.

(One of the original requests that got lost in the dark... #476)

**Comment by jasonnutter** [comment]
@hansmaad We have not had a chance to prioritize this, unfortunately. You are welcome to open PR, however!

**Comment by andreaslarssen** [comment]
My issue (#4932) was closed in favor of this one. @bmahall answered that one with `I am currently looking into this , thanks for your patience!`, but this issue hasn't moved in almost a year.

**Comment by hansmaad** [comment]
It would be great if the interceptor would be more open for extension in general. Every little attempt for customization usually means that we have to copy the whole thing and do changes instead of extending it.

**Comment by bytelabsco** [comment]
I opened a PR almost a year ago implementing the changes that @hansmaad mentioned, but it seems to have been lost in the shuffle.

**Comment by jdussouillez** [comment]
Any update on this?
